tempodb/backend/azure/azure_helpers.go

```diff
@@ -196,3 +242,32 @@ func getServicePrincipalToken(cfg *Config) (*adal.ServicePrincipalToken, error)
 
 	return msiConfig.ServicePrincipalToken()
 }
+
+func servicePrincipalTokenFromFederatedToken(environment string, resource string, newOAuthConfigFunc func(activeDirectoryEndpoint, tenantID string) (*adal.OAuthConfig, error), newServicePrincipalTokenFromFederatedTokenFunc func(oauthConfig adal.OAuthConfig, clientID string, jwt string, resource string, callbacks ...adal.TokenRefreshCallback) (*adal.ServicePrincipalToken, error)) (*adal.ServicePrincipalToken, error) {
```

Comment:
as suggested above, let's change this to take the `authFunctions` struct. Alternatively we can create function types to clean up this function signature. Something like:

```
type oAuthConfig func(activeDirectoryEndpoint, tenantID string) (*adal.OAuthConfig, error)
```

Reply:
Done in https://github.com/grafana/tempo/pull/2195/commits/83b5d32b2d091cd8646ef047d99c26d802e6d8ce